packages/okta-react/src/AuthService.js

```diff
@@ -43,45 +41,123 @@ export default class Auth {
+
+      for (let token of tokens) {
+        if (token.idToken) {
+          this._oktaAuth.tokenManager.add('idToken', token);
+        } else if (token.accessToken) {
+          this._oktaAuth.tokenManager.add('accessToken', token);
+        }
+      }
+      const authState = await this.updateAuthState();
+      if(authState.isAuthenticated) { 
+        const location = this.getFromUri();
+        window.location.assign(location);
+      }
+      this._pending.handleAuthentication = null;
+    } catch(error) { 
+      this._pending.handleAuthentication = null;
+      this.emitAuthState({ 
+        isAuthenticated: false,
+        error,
+        idToken: null,
+        accessToken: null,
+      });
+    } 
+    return;
+  }
+
+  clearAuthState(state={}) { 
+    this.emitAuthState({ ...AuthService.DEFAULT_STATE, ...state });
+    return;
+  }
+
+  emitAuthState(state) { 
+    this._authState = state;
+    this.emit('authStateChange', this.getAuthState());
+    return;
+  }
+
+  getAuthState() { 
+    return this._authState;
   }
 
-  async isAuthenticated() {
-    // Support a user-provided method to check authentication
-    if (this._config.isAuthenticated) {
-      return (this._config.isAuthenticated)();
+  async updateAuthState() {
+    // avoid concurrent updates
+    if( this._pending.authStateUpdate ) { 
+      return this._pending.authStateUpdate.promise;
     }
 
-    // If there could be tokens in the url
-    if (location && location.hash && containsAuthTokens.test(location.hash)) return null;
+    // create a promise to return in case of multiple parallel requests
+    this._pending.authStateUpdate = {};
+    this._pending.authStateUpdate.promise  = new Promise( (resolve) => {
+      // Promise can only resolve any error is in the resolve value
+      // and uncaught exceptions make Front SDKs angry
+      this._pending.authStateUpdate.resolve = resolve;
+    });
+    // copy to return after emitAuthState has cleared the pending object
+    const authStateUpdate = this._pending.authStateUpdate;
 
-    // Return true if either the access or id token exist in client storage
-    return !!(await this.getAccessToken()) || !!(await this.getIdToken());
+    try { 
+      const accessToken = await this.getAccessToken();
+      const idToken = await this.getIdToken();
+
+      // Use external check, or default to isAuthenticated if either the access or id token exist
+      const isAuthenticated = this._config.isAuthenticated ? await this._config.isAuthenticated() : !! ( accessToken || idToken );
+
+
+      this._pending.authStateUpdate = null;
+      this.emitAuthState({ 
+        isAuthenticated,
+        idToken,
+        accessToken,
+      });
+    } catch (error) { 
+      this._pending.authStateUpdate = null;
+      this.emitAuthState({ 
+        isAuthenticated: false,
+        error,
+        idToken: null,
+        accessToken: null,
+      });
+    } 
+    authStateUpdate.resolve();
```

Comment:
I think `resolve()` needs to be called with a value, otherwise the promise will resolve `undefined`. This will break `handleAuthentication` which is expecting a value to be returned: https://github.com/okta/okta-oidc-js/pull/677/files#diff-8a51cbb87b5cfd7e092f0410797f00e9R87
`handleAuthentication` should either read from `authState.getAuthState()` (and this promise is defined as always returning undefined), or the promise should resolve to the authState.

Reply:
Ah, I missed a place to update.

I was removing all the numerous places authState was returned to make it clear to consult the one source of truth (authService.getAuthState()), but I missed this update.

Once more unto the breach.